### ISSUE TITLE
downloadProgress block in GET is not called under AFNetworking 3.0

### ISSUE THREAD

**Issue by kunass2** [AFNetworking/AFNetworking#3270]
I have created a subclass of `AFHTTPSessionManager`:

    class TBNetworkClient: AFHTTPSessionManager {
        func wordsWithCompletionBlock(completionBlock: TBErrorHandler) {
        
            GET("words", parameters: nil, progress: { progress in
            
                print("----------") //this is not called at all, why?
            
                }, success: { sessionDataTask, response in

                    print((sessionDataTask.response as? NSHTTPURLResponse)?.allHeaderFields)
                    completionBlock(nil) //success is called

                }) { sessionDataTask, error in

                    completionBlock(error)
            }
        }
    }

Marked line is not called at all, why?

The output in success is following:

>Optional([Server: Apache, Connection: Keep-Alive, Content-Type: text/html, Date: Fri, 08 Jan 2016 08:58:18 GMT, Content-Encoding: gzip, Keep-Alive: timeout=1, max=100, Content-Length: 5663, Vary: Accept-Encoding,User-Agent])

**environment**: Xcode 7 and iOS 9.0, latest version of AFNetworking from CocoaPods

[Progress block with GET and AFNetworking 3.0 is not called](http://stackoverflow.com/q/34662531/2725435)

**Comment by kcharwood** [comment]
NSURLSessionTasks only report progress if the `Content-Length` header is present, which it appears to be there based on your response.

Could you include the URL you are trying to hit? We have unit tests around this feature, so you can see that it _is_ working. I would assume something is going on with the response from the server, and start looking there.

**Comment by kunass2** [comment]
Of course: http://www.taboo.blue-world.pl/api/words

but you must additionally set two headers to get the response.

    setValue("normal", forHTTPHeaderField: "X-Type")
    setValue("pl", forHTTPHeaderField: "X-Language")

**Comment by emilyZhouwm** [comment]
iOS8.0, 9.0 is OK, but progress block not called at all，iOS7.0

    AFURLSessionManager *manager = [EHDownloadNetwork sharedInstance].manager;
    downloadTask.task = [manager downloadTaskWithRequest:request
              progress:^(NSProgress * _Nonnull downloadProgress) {
                  if (progressBlock) {
                      dispatch_async(dispatch_get_main_queue(), ^{
                          progressBlock(downloadProgress.fractionCompleted);
                      });
                  }

**Comment by kcharwood** [comment]
@emilyZhouwm 

Can you provide more information about the response the server is sending? AFNetworking hooks into the NSURLSession progress update properties, which rely on a response to properly format the expected amount of bytes to be returned by the response.

**Comment by kunass2** [comment]
@kcharwood do you know more about how to solve this issue?

**Comment by kcharwood** [comment]
The tl;dr; version is AFNetworking only gets progress callbacks if NSURLSession gets progress callbacks. NSURLSession relies on properly formatted response headers to help determine how many bytes are left to be downloaded. AFNetworking is simply [wrapping KVO](https://github.com/AFNetworking/AFNetworking/blob/master/AFNetworking/AFURLSessionManager.m#L187) of the session data task.

**Comment by kunass2** [comment]
So it is impossible to make progress working with simple GET?

**Comment by kcharwood** [comment]
No. Progress does work with a simple GET. There are many tests in the test suite showing that. But it does require the server to return proper response headers to support it at the NSURLSession level.

**Comment by kunass2** [comment]
So, could you help me investigate what is wrong with my response?

**Comment by kcharwood** [comment]
I'd start by looking at the test suite, and checking that out. Those tests hit httpbin.org, which is public and you can see all the headers returned. I'd then compare and contrast those with the headers you get back, and see if we can pick anything out.

**Comment by emilyZhouwm** [comment]
NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url]];
    
    // 检查文件是否已经下载了一部分
    unsigned long long downloadedBytes = 0;
    NSString *tempPath = [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0] stringByAppendingString:destPath];
    if ([[NSFileManager defaultManager] fileExistsAtPath:tempPath]) {
        // 获取已下载的文件长度
        downloadedBytes = [self fileSizeForPath:tempPath];
        if (downloadedBytes > 0) {
            NSString *requestRange = [NSString stringWithFormat:@"bytes=%llu-", downloadedBytes];
            [request setValue:requestRange forHTTPHeaderField:@"Range"];
        }
    }

    // 不使用缓存，避免断点续传出现问题
    [[NSURLCache sharedURLCache] removeCachedResponseForRequest:request];
    
    NSLog(@"\n===========download===========\n%@", url);

    EHDownloadTask *downloadTask = [[EHDownloadTask alloc] init];
    AFURLSessionManager *manager = [EHDownloadNetwork sharedInstance].manager;
    downloadTask.task = [manager downloadTaskWithRequest:request
              progress:^(NSProgress * _Nonnull downloadProgress) {
                  if (progressBlock) {
                      dispatch_async(dispatch_get_main_queue(), ^{
                          progressBlock(downloadProgress.fractionCompleted);
                      });
                  }
              } 
。。。。。。。

**Comment by emilyZhouwm** [comment]
没有任何log

**Comment by emilyZhouwm** [comment]
![2a9769c8-b03d-40e1-868d-4e56e60885fa](https://cloud.githubusercontent.com/assets/6175955/12344885/3bfebe32-bb7d-11e5-961b-9fe966bc7470.png)
![aa2a9d4a-e250-498d-9538-e1c2c38cb904](https://cloud.githubusercontent.com/assets/6175955/12344886/3f17ff66-bb7d-11e5-86ae-1d1fe017f1c4.png)

**Comment by kunass2** [comment]
Ok, So I need to wait then... no problem. If I could help you somehow, let me know. I think that it is quite possible that some headers are wrong or do not exists at all on my server response.

**Comment by kcharwood** [comment]
I just pushed in a patch from @emilyZhouwm that fixes this for download tasks on iOS 7.

At this point, I'm not aware of any other issues. If someone can provide a clear example of HTTP response headers for a request where progress is not working, we can move forward on debugging any other progress related issues.

:beers:

**Comment by kunass2** [comment]
Will it be available through CocoaPods also?

**Comment by kcharwood** [comment]
It will be available in the 3.1.0 release of AFNetworking, and is currently available in master.

**Comment by AWDGroup** [comment]
@kcharwood 
I have the same issue and I'm using the latest version of AFNetworking 3.1.0 and iOS 9.3,
downloadProgress does not get called under downloadTaskWithRequest:

```
NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:configuration];

NSURL *URL = [NSURL URLWithString:@"https://s3.amazonaws.com/awdtest/fullzip.pdf"];
NSURLRequest *request1 = [NSURLRequest requestWithURL:URL];

NSURLSessionDownloadTask *downloadTask1 = [manager downloadTaskWithRequest:request1 progress:^(NSProgress * _Nonnull downloadProgress)
{
    NSLog(@"Progress: %f", downloadProgress.fractionCompleted);
} destination:^NSURL *(NSURL *targetPath, NSURLResponse *response)
{
    NSURL *documentsDirectoryURL = [[NSFileManager defaultManager] URLForDirectory:NSDocumentDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:NO error:nil];
    return [documentsDirectoryURL URLByAppendingPathComponent:[response suggestedFilename]];
} completionHandler:^(NSURLResponse *response, NSURL *filePath, NSError *error)
{
    NSLog(@"File downloaded to: %@", filePath);
}];
[downloadTask1 resume];
```

Response headers:

```
HTTP/1.1 200 OK
x-amz-id-2: CW06YcgIycOHZQy8bCJrT3aNfhatM9pty1mOjgYHumjCxRmNAQ+jhJHRQwl7mDIaQeTHI0fyrnU=
x-amz-request-id: 105E010DECC91897
Date: Sat, 14 May 2016 09:26:38 GMT
Content-Encoding: gzip
Last-Modified: Sat, 14 May 2016 09:21:29 GMT
ETag: "88bbe0b318bf11dd56a31176d3384e78"
Accept-Ranges: bytes
Content-Type: application/pdf
Content-Length: 1243325
Server: AmazonS3
```


it works if I use this file which is not gzipped: https://s3.amazonaws.com/awdtest/full.pdf

Response headers:

```
HTTP/1.1 200 OK
x-amz-id-2: gFOVfhheaMdeJOBb+7H8oaXjLLeOqlQl616XnYx6C2Gj7PBVLKZ9kMIN2fJOrGBcSgQ/7nbQOc0=
x-amz-request-id: 26669D9B576E300A
Date: Sat, 14 May 2016 09:54:58 GMT
Last-Modified: Fri, 16 May 2014 05:42:35 GMT
ETag: "6f16d7e09023ce8b50fd67abba7825c4"
Accept-Ranges: bytes
Content-Type: application/pdf
Content-Length: 1411131
Server: AmazonS3
```
